cpu/sam3: uart driver enhancements

haukepetersen opened the pull request:
Follow up on #6568

- enabled driver to run in TX only mode
- make use of gpio_init_mux() for pin intialization
- simplyfied pin configuration
- adapted pin configuartion for arduino-due and udoo

aabadie commented:
Tested `examples/default` on arduino-due : the shell is broken. It works with master.

haukepetersen commented:
rebased.

Just tested the `arduino-due` with the `default` example again, works just fine here. Could you check again, if the rebase had anything to do with it? Otherwise a toolchain issue? I am using `arm-none-eabi-gcc (GNU Tools for ARM Embedded Processors) 5.4.1 20160919 (release)`.

aabadie commented:
My compiler version outputs this:
```
arm-none-eabi-gcc (15:4.9.3+svn231177-1) 4.9.3 20150529 (prerelease)
```
I'll test again tomorrow.

aabadie commented on cpu/sam3/periph/uart.c:
```diff
@@ -49,25 +50,30 @@ int uart_init(uart_t uart, uint32_t baudrate, uart_rx_cb_t rx_cb, void *arg)
+    gpio_init(uart_config[uart].tx_pin, GPIO_OUT);
+    gpio_init_mux(uart_config[uart].tx_pin, uart_config[uart].mux);
+    if (rx_cb) {
+        gpio_init(uart_config[uart].rx_pin, GPIO_IN);
```
It should be `GPIO_IN_PD` here. Otherwise it doesn't work for me.

haukepetersen replied:
Hmm, a pull-down on the RX pin? That sounds strange... Is there some reference to this in the datasheet?

aabadie replied:
I can't explain it either. I also tried with `GPIO_IN_PU` and it also works. There's nothing about this in the datasheet.

aethaniel commented on boards/arduino-due/include/periph_conf.h:
```diff
@@ -68,11 +68,38 @@ static const timer_conf_t timer_config[] = {
+    },
+    {
+        .dev    = (Uart *)USART3,
+        .rx_pin = GPIO_PIN(PD, 4),
```
rx and tx are switched. see SAM3X/A datasheet page 44

aethaniel commented on boards/udoo/include/periph_conf.h:
```diff
@@ -66,11 +66,38 @@ static const timer_conf_t timer_config[] = {
+    },
+    {
+        .dev    = (Uart *)USART3,
+        .rx_pin = GPIO_PIN(PD, 4),
```
ditto

haukepetersen commented:
@aethaniel thanks for spotting the pin-hickup! Fixed, rebased, and squashed.

aabadie commented:
> Tested examples/default on arduino-due : the shell is broken

Ok, the problem I had has probably something to do with my compiler version. I test again asap.

haukepetersen commented:
awesome

aabadie commented:
> the problem I had has probably something to do with my compiler version

@haukepetersen, just re-tested and still have the issue. Compiler version: 6.3.1 20170215

haukepetersen commented:
rebased.

@aabadie i removed the `gpio_init()` calls all together, as there is actually nothing in the ref manual that tells us to set the pins as input or output. Simply calling `gpio_init_mux()` should be sufficient. Could you test this again? Thx

haukepetersen commented:
awesome, done.

haukepetersen commented:
is something wrong with Murdock? Shouldn't it run automatically here?

aabadie commented:
Indeed, something doesn't work with Murdock

aabadie commented:
All green, go!
